File: rust/checks/dependency-boundaries/check.mjs
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import { isAbsolute, relative, resolve, sep } from "node:path";

const requestedRoot = resolve(process.argv[2]);
const metadata = JSON.parse(
    execFileSync(
        "cargo",
        [
            "metadata",
            "--no-deps",
            "--format-version",
            "1",
            "--locked",
            "--offline",
        ],
        { cwd: resolve(requestedRoot, "rust"), encoding: "utf8" },
    ),
);
const root = resolve(metadata.workspace_root, "..");
const bindings = resolve(root, "rust/bindings");
const framework = /^(?:wasm-bindgen(?:-.*)?|js-sys|web-sys|serde-wasm-bindgen|tsify(?:-.*)?|flutter_rust_bridge(?:_.*)?|uniffi(?:_.*)?|napi(?:-.*)?)$/;
const pathFromRoot = (path) => relative(root, path).split(sep).join("/");
const domain = metadata.packages.filter((pkg) =>
    pathFromRoot(pkg.manifest_path).startsWith("rust/crates/"),
);

const httpOwners = new Set([
    "rust/crates/core/Cargo.toml",
    "rust/crates/assets/Cargo.toml",
    "rust/crates/location-dataset/Cargo.toml",
]);

for (const pkg of metadata.packages) {
    const manifest = pathFromRoot(pkg.manifest_path);
    if (httpOwners.has(manifest)) continue;
    for (const dependency of pkg.dependencies) {
        if (dependency.name !== "reqwest" || dependency.kind === "dev") continue;
        console.error(
            `${manifest}: use ente-core::http instead of a production reqwest dependency`,
        );
        process.exitCode = 1;
    }
}

if (!domain.length) {
    console.error(`${resolve(root, "rust/crates")}: found no domain crates`);
    process.exitCode = 1;
}

for (const pkg of domain) {
    const manifest = pathFromRoot(pkg.manifest_path);
    for (const dependency of pkg.dependencies) {
        if (!framework.test(dependency.name) && !inside(bindings, dependency.path)) continue;
        console.error(
            `${manifest}: dependency ${JSON.stringify(dependency.name)} crosses the Rust binding boundary`,
        );
        process.exitCode = 1;
    }
}

const registered = new Set(domain.map((pkg) => pathFromRoot(pkg.manifest_path)));
for (const manifest of git(
    "ls-files",
    "--cached",
    "--others",
    "--exclude-standard",
    "-z",
    "--",
    "rust/crates/**/Cargo.toml",
)
    .split("\0")
    .filter((path) => path && existsSync(resolve(root, path)))) {
    if (registered.has(manifest)) continue;
    console.error(`${manifest}: domain crate is not registered in the Cargo workspace`);
    process.exitCode = 1;
}

function git(...args) {
    return execFileSync("git", args, { cwd: root, encoding: "utf8" });
}

function inside(directory, path) {
    if (!path) return false;
    const local = relative(directory, path);
    return local !== ".." && !local.startsWith(`..${sep}`) && !isAbsolute(local);
}

File: rust/checks/cargo-versions/check.py
import re
import subprocess
import sys
import tomllib
from pathlib import Path

root = Path(sys.argv[1]).resolve()
sections = "dependencies", "dev-dependencies", "build-dependencies"
complete = re.compile(
    r"(?:\^|~|=|>=|>|<=|<)?\s*[0-9]+\.[0-9]+\.[0-9]+"
    r"(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?"
    r"(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?"
)
manifest_paths = subprocess.check_output(
    [
        "git",
        "ls-files",
        "--cached",
        "--others",
        "--exclude-standard",
        "-z",
        "--",
        "rust/Cargo.toml",
        "rust/**/Cargo.toml",
    ],
    cwd=root,
    text=True,
).split("\0")
manifests = sorted(
    root / path for path in manifest_paths if path and (root / path).is_file()
)
if not manifests:
    raise SystemExit(f"{root}/rust: found no Cargo manifests")
failed = False

for manifest in manifests:
    path = manifest.relative_to(root).as_posix()
    cargo = tomllib.loads(manifest.read_text())
    tables = [(section, cargo.get(section, {})) for section in sections]
    tables += [
        (f"target.{target}.{section}", config.get(section, {}))
        for target, config in cargo.get("target", {}).items()
        for section in sections
    ]
    tables.append(
        ("workspace.dependencies", cargo.get("workspace", {}).get("dependencies", {}))
    )
    for table, dependencies in tables:
        for name, declaration in dependencies.items():
            requirement = (
                declaration if isinstance(declaration, str) else declaration.get("version")
            )
            if requirement is None:
                continue
            if all(complete.fullmatch(part.strip()) for part in requirement.split(",")):
                continue
            print(
                f"{path}: {table}.{name} has incomplete version requirement {requirement!r}",
                file=sys.stderr,
            )
            failed = True

sys.exit(1 if failed else 0)
